docs/sphinx/developers/wsi.rst

```diff
@@ -0,0 +1,63 @@
+will load from the full resolution image in the first pyramid stored in the file.  Each resolution of
+each pyramid is stored as a separate series, and can be accessed by calling
+:javadoc:`setSeries <loci/formats/IFormatReader.html#setSeries-int->` prior to retrieving pixel data.
+
+There are additional API methods that can be used to make pyramids easier to work with.  These can be enabled by calling
+:javadoc:`setFlattenedResolutions(true) <loci/formats/IFormatReader.html#setFlattenedResolutions-boolean->`
+prior to :javadoc:`setId <loci/formats/IFormatHandler.html#setId-java.lang.String->`.
+
+After ``setUnflattenedResolutions(true)``, each series represents an entire image pyramid and not
+just a single resolution.  Calling ``setSeries(...)`` then skips over
+all other resolutions in the same pyramid, to either the next pyramid (if
+multiple pyramids are stored), or the thumbnail or barcode image (if present).
+To access the smaller resoultions in the pyramid, use the :javadoc:`getResolutionCount() <loci/formats/IFormatReader.html#getResolutionCount-->`
+and :javadoc:`setResolution(int) <loci/formats/IFormatReader.html#setResolution-int->` methods.
+
+Most formats only store one pyramid per fileset, but some (e.g. :doc:`/formats/cellsens-vsi`)
+allow multiple pyramids.  Almost all formats allow a thumbnail, slide
+overview, and/or slide barcode image.  Bio-Formats always stores these images
+as separate series, after all of the pyramids.  Be careful to check the pixel
+type for the extra images, as the type and channel count will often differ
+from that of the pyramid(s).
+
+For an example of how to use the pyramid resolution API, see
+:source:`SubResolutionExample.java <components/formats-gpl/utils/SubResolutionExample.java>`.
+
+Bio-Formats also provides some visibility into how the tiles are stored via the
+:javadoc:`getOptimalTileWidth() <loci/formats/IFormatReader.html#getOptimalTileWidth-->` and
+:javadoc:`getOptimalTileHeight() <loci/formats/IFormatReader.html#getOptimalTileHeight-->` methods.
+This is a suggestion of the size of tiles to be passed to
+:javadoc:`openBytes(int, byte[], int, int, int, int) <loci/formats/IFormatReader.html#openBytes-int-byte:A-int-int-int-int->`,
+in order to minimize the number of tile decompressions.  In most cases, and especially for
+the largest resolution, the whole image can't be loaded at once.  The amount
+of memory allocated is not a factor in being able to load the whole image, as
+no more than 2GB of pixel data can be stored in a single byte array and most
+full resolution images will exceed this limit.
+
+Internal OMERO pyramid format
+=============================
+
+For files that contain very large images and are not in a format that supports pyramids, OMERO will generate its own
+image pyramid to improve visualization performance.  Bio-Formats can read these generated pyramids, but cannot
+currently write them outside of OMERO.
```

Comment:
Can we actually read these pyramid files with the subresolutions using vanilla Bio-Formats? I was under the impression that https://github.com/openmicroscopy/openmicroscopy/blob/v5.3.3/components/romio/src/ome/io/bioformats/OmeroPixelsPyramidReader.java was the reader used for these pyramids.